File: ios-knitstudio/KnitStudio/Import/WrittenPatternParser.swift
import Foundation

/// One instruction inside a row, and what it does to the stitch count.
struct StitchOperation: Equatable {
    var text: String
    var consumes: Int
    var produces: Int
}

/// One row of a written pattern, after parsing.
struct ParsedRow: Identifiable, Equatable {
    var id: UUID = UUID()
    var label: String
    var side: String?
    var text: String
    var startingStitches: Int?
    var endingStitches: Int?
    var issues: [String] = []

    var isVerified: Bool { endingStitches != nil && issues.isEmpty }
}

/// The result of importing a written pattern.
struct ParsedPattern: Equatable {
    var title: String?
    var gauge: Gauge?
    var gaugeNote: String?
    var castOnStitches: Int?
    var yarnNotes: [String] = []
    var needleNotes: [String] = []
    var rows: [ParsedRow] = []
    var unknownTerms: [String] = []
    var issues: [String] = []

    var verifiedRowCount: Int { rows.filter(\.isVerified).count }
}

/// Reads a written knitting pattern and works out what each row does to the
/// stitch count — which is how you find the row where a pattern (or a
/// transcription of it) stops adding up.
enum WrittenPatternParser {

    // MARK: - Entry point

    static func parse(_ raw: String) -> ParsedPattern {
        var result = ParsedPattern()
        let lines = raw
            .replacingOccurrences(of: "\r\n", with: "\n")
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }

        result.title = lines.first { !$0.isEmpty && $0.count < 80 }

        for line in lines where !line.isEmpty {
            let lower = line.lowercased()

            if result.gauge == nil, lower.contains("gauge") || lower.contains("tension")
                || (lower.contains("sts") && (lower.contains("row") || lower.contains("rnd"))) {
                if let parsed = parseGauge(line) {
                    result.gauge = parsed
                    result.gaugeNote = line
                }
            }
            if lower.contains("yarn") || lower.contains("skein") || lower.contains("ball")
                || lower.contains("yardage") || lower.contains("metres") || lower.contains("meters") {
                result.yarnNotes.append(line)
            }
            if lower.contains("needle") || (lower.contains("mm") && lower.contains("circular")) {
                result.needleNotes.append(line)
            }
            if result.castOnStitches == nil,
               let count = parseCastOn(line) {
                result.castOnStitches = count
            }
            if let row = parseRowLine(line) {
                result.rows.append(row)
            }
        }

        // Walk the rows, threading the stitch count through.
        var running = result.castOnStitches
        var unknown = Set<String>()
        for index in result.rows.indices {
            result.rows[index].startingStitches = running
            let outcome = evaluate(result.rows[index].text, incoming: running)
            result.rows[index].endingStitches = outcome.produced
            result.rows[index].issues = outcome.issues
            outcome.unknown.forEach { unknown.insert($0) }
            running = outcome.produced ?? running
        }
        result.unknownTerms = unknown.sorted()

        if result.rows.isEmpty {
            result.issues.append(
                "No numbered rows found. This importer reads patterns written as "
                + "‘Row 1 (RS): k2, p2…’ — a scanned image of a chart will not parse.")
        }
        if result.castOnStitches == nil, !result.rows.isEmpty {
            result.issues.append(
                "No cast-on count found, so stitch counts could not be checked. "
                + "Add a line like ‘Cast on 96 sts’ and import again.")
        }
        return result
    }

    // MARK: - Header lines

    static func parseGauge(_ line: String) -> Gauge? {
        guard let match = capture(
            #"(\d+(?:[.,]\d+)?)\s*(?:sts?|stitches)\D{0,40}?(\d+(?:[.,]\d+)?)\s*(?:rows?|rnds?|rounds?)"#,
            in: line
        ), match.count >= 3,
            let stitches = number(match[1]),
            let rows = number(match[2])
        else { return nil }

        // Work out whether the gauge is quoted over 10 cm or 4 inches.
        var window = 10.0
        if let unit = capture(#"(\d+(?:[.,]\d+)?)\s*(cm|centimetre|centimeter|in\b|inch|inches|")"#, in: line),
           unit.count >= 3, let value = number(unit[1]) {
            let isInches = (unit[2] ?? "").lowercased().hasPrefix("in") || (unit[2] ?? "") == "\""
            window = isInches ? value * 2.54 : value
        } else if line.lowercased().contains("4\"") || line.lowercased().contains("4 in") {
            window = 10.16
        }
        guard window > 0 else { return nil }
        return Gauge.fromSwatch(stitches: stitches, rows: rows, overCm: window)
    }

    static func parseCastOn(_ line: String) -> Int? {
        // The \b stops "co" matching inside another word.
        guard let match = capture(#"(?:cast on|\bco)\s+(\d+)\s*(?:sts?|stitches)?"#, in: line),
              match.count >= 2, let value = number(match[1]) else { return nil }
        return Int(value)
    }

    static func parseRowLine(_ line: String) -> ParsedRow? {
        guard let match = capture(
            #"^(rows?|rnds?|rounds?)\s*([\d\s,\-–and]+?)\s*(?:\((rs|ws)\))?\s*[:.]\s*(.+)$"#,
            in: line
        ), match.count >= 5 else { return nil }

        let keyword = (match[1] ?? "Row").capitalized
        let numbers = (match[2] ?? "").trimmingCharacters(in: .whitespaces)
        let side = match[3]?.uppercased()
        let body = (match[4] ?? "").trimmingCharacters(in: .whitespaces)
        guard !body.isEmpty else { return nil }

        return ParsedRow(
            label: "\(keyword) \(numbers)",
            side: side,
            text: body)
    }

    // MARK: - Row evaluation

    struct Outcome {
        var produced: Int?
        var issues: [String] = []
        var unknown: [String] = []
    }

    /// Works out the stitch count a row ends on, given what it started with.
    static func evaluate(_ text: String, incoming: Int?) -> Outcome {
        var outcome = Outcome()
        let normalised = normalise(text)

        // Pull the pattern's own stated count out of a trailing "— 96 sts".
        let stated = statedCount(in: normalised)

        guard let incoming else {
            outcome.produced = stated
            return outcome
        }

        let expanded = expandBrackets(normalised)

        // Split around a star repeat, if there is one.
        if let star = splitStarRepeat(expanded) {
            let prefix = tally(star.prefix, collecting: &outcome)
            let body = tally(star.body, collecting: &outcome)
            let suffix = tally(star.suffix, collecting: &outcome)

            guard body.consumes > 0 else {
                outcome.issues.append("The repeated section does not use any stitches, so it cannot be counted.")
                outcome.produced = stated
                return outcome
            }
            // "to last N sts" describes the same stitches the suffix works, so
            // it is a consistency check, not an extra deduction.
            if star.leaveStitches > 0, suffix.consumes != star.leaveStitches {
                outcome.issues.append(
                    "‘to last \(star.leaveStitches) sts’ but the instructions after it use "
                    + "\(suffix.consumes) sts.")
            }
            let available = incoming - prefix.consumes - suffix.consumes
            guard available >= 0 else {
                outcome.issues.append(
                    "This row needs more stitches than the \(incoming) on the needle.")
                outcome.produced = stated
                return outcome
            }
            let repeats = available / body.consumes
            if available % body.consumes != 0 {
                outcome.issues.append(
                    "The repeat does not divide evenly: \(available) sts left for a "
                    + "\(body.consumes)-st repeat leaves \(available % body.consumes) over.")
            }
            let produced = prefix.produces + repeats * body.produces + suffix.produces
            outcome.produced = produced
            check(produced: produced, stated: stated, into: &outcome)
            return outcome
        }

        // No repeat: a straight tally, with "to end" soaking up the remainder.
        let flexible = flexibleConsumption(expanded)
        let fixed = tally(flexible.tokens, collecting: &outcome)
        if flexible.hasToEnd {
            // The instructions after "to last N sts" are exactly those N sts —
            // anything before the marker is not part of that count.
            var afterOutcome = Outcome()
            let after = tally(flexible.after, collecting: &afterOutcome)
            if flexible.leaveStitches > 0, after.consumes != flexible.leaveStitches {
                outcome.issues.append(
                    "‘to last \(flexible.leaveStitches) sts’ but the instructions after it use "
                    + "\(after.consumes) sts.")
            }
            let available = incoming - fixed.consumes
            guard available >= 0 else {
                outcome.issues.append(
                    "This row needs more stitches than the \(incoming) on the needle.")
                outcome.produced = stated
                return outcome
            }
            let produced = fixed.produces + available
            outcome.produced = produced
            check(produced: produced, stated: stated, into: &outcome)
            return outcome
        }

        if fixed.consumes != incoming {
            outcome.issues.append(
                "This row uses \(fixed.consumes) sts but \(incoming) were on the needle.")
        }
        outcome.produced = fixed.produces
        check(produced: fixed.produces, stated: stated, into: &outcome)
        return outcome
    }

    static func check(produced: Int, stated: Int?, into outcome: inout Outcome) {
        guard let stated, stated != produced else { return }
        outcome.issues.append(
            "The pattern says \(stated) sts at the end of this row; the instructions give \(produced).")
    }

    // MARK: - Tokenising

    struct Tally {
        var consumes: Int = 0
        var produces: Int = 0
    }

    static func tally(_ tokens: [String], collecting outcome: inout Outcome) -> Tally {
        var result = Tally()
        for token in tokens {
            guard let operation = interpret(token) else {
                if !token.isEmpty { outcome.unknown.append(token) }
                continue
            }
            result.consumes += operation.consumes
            result.produces += operation.produces
        }
        return result
    }

    /// Turns one instruction into its stitch arithmetic.
    static func interpret(_ token: String) -> StitchOperation? {
        let text = token.trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty else { return nil }

        // Instructions that move markers or turn the work change nothing.
        for noop in ["pm", "sm", "place marker", "slip marker", "remove marker", "turn", "rm"]
        where text == noop || text.hasPrefix(noop + " ") {
            return StitchOperation(text: text, consumes: 0, produces: 0)
        }

        // Multi-stitch operations, longest first so "k2tog" beats "k2".
        let fixedOperations: [(String, Int, Int)] = [
            ("sl2-k1-p2sso", 3, 1), ("sl2 k1 p2sso", 3, 1), ("s2kp", 3, 1), ("sk2p", 3, 1),
            ("cdd", 3, 1), ("k3tog", 3, 1), ("p3tog", 3, 1),
            ("k2tog tbl", 2, 1), ("p2tog tbl", 2, 1),
            ("k2tog", 2, 1), ("p2tog", 2, 1), ("ssk", 2, 1), ("ssp", 2, 1),
            ("kfb", 1, 2), ("pfb", 1, 2),
            ("m1lp", 0, 1), ("m1rp", 0, 1), ("m1l", 0, 1), ("m1r", 0, 1), ("m1p", 0, 1), ("m1", 0, 1),
            ("yo", 0, 1), ("yon", 0, 1), ("yfwd", 0, 1), ("yrn", 0, 1),
        ]
        for (name, consumes, produces) in fixedOperations where text == name || text.hasPrefix(name + " ") {
            return StitchOperation(text: text, consumes: consumes, produces: produces)
        }

        // "k12", "p3", "sl 2", "knit 5", and bare "k" / "p".
        if let match = capture(#"^(k|p|knit|purl|sl|slip)\s*(\d+)?(?:\s*sts?)?(?:\s*tbl)?$"#, in: text),
           match.count >= 2 {
            let count = match[2].flatMap { Int($0) } ?? 1
            return StitchOperation(text: text, consumes: count, produces: count)
        }

        // "bind off 8 sts" / "cast on 12 sts".
        if let match = capture(#"^(?:bind off|bo|cast off)\s*(\d+)"#, in: text), match.count >= 2,
           let count = Int(match[1] ?? "") {
            return StitchOperation(text: text, consumes: count, produces: 0)
        }
        if let match = capture(#"^(?:cast on|co)\s*(\d+)"#, in: text), match.count >= 2,
           let count = Int(match[1] ?? "") {
            return StitchOperation(text: text, consumes: 0, produces: count)
        }

        return nil
    }

    // MARK: - Structure

    static func normalise(_ text: String) -> String {
        text.lowercased()
            .replacingOccurrences(of: "–", with: "-")
            .replacingOccurrences(of: "—", with: "-")
            .replacingOccurrences(of: "×", with: "x")
            .replacingOccurrences(of: "  ", with: " ")
            .trimmingCharacters(in: .whitespaces)
    }

    /// Reads a trailing "(96 sts)" or "- 96 sts" the pattern states for itself.
    static func statedCount(in text: String) -> Int? {
        guard let match = capture(#"[\(\-]\s*(\d+)\s*sts?\s*\)?\s*$"#, in: text),
              match.count >= 2, let value = Int(match[1] ?? "") else { return nil }
        return value
    }

    /// Expands "[k2, p2] 3 times" and "(k1, yo) x4" into plain instructions.
    static func expandBrackets(_ text: String) -> String {
        var working = text
        let pattern = #"[\[\(]([^\]\)]+)[\]\)]\s*(?:x\s*)?(\d+)\s*(?:times?)?"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return working
        }
        // Innermost-first, and bounded so a malformed pattern cannot loop forever.
        for _ in 0 ..< 8 {
            let range = NSRange(working.startIndex ..< working.endIndex, in: working)
            guard let match = regex.firstMatch(in: working, options: [], range: range),
                  let whole = Range(match.range, in: working),
                  let bodyRange = Range(match.range(at: 1), in: working),
                  let countRange = Range(match.range(at: 2), in: working),
                  let count = Int(working[countRange]), count > 0, count < 400
            else { break }

            let body = String(working[bodyRange]).trimmingCharacters(in: .whitespaces)
            let expansion = Array(repeating: body, count: count).joined(separator: ", ")
            working.replaceSubrange(whole, with: expansion)
        }
        return working
    }

    struct StarRepeat {
        var prefix: [String]
        var body: [String]
        var suffix: [String]
        var leaveStitches: Int
    }

    /// Splits "k3, *p2, k2; rep from * to last 3 sts, k3" into its three parts.
    static func splitStarRepeat(_ text: String) -> StarRepeat? {
        guard let starIndex = text.firstIndex(of: "*") else { return nil }
        let prefixText = String(text[text.startIndex ..< starIndex])
        let rest = String(text[text.index(after: starIndex)...])

        guard let repeatRange = rest.range(
            of: #"rep(eat)?\s+from\s*\*"#,
            options: [.regularExpression, .caseInsensitive]
        ) else { return nil }

        let bodyText = String(rest[rest.startIndex ..< repeatRange.lowerBound])
        let tail = String(rest[repeatRange.upperBound...])

        var leave = 0
        var suffixText = tail
        if let match = capture(#"to\s+last\s+(\d+)\s*sts?"#, in: tail), match.count >= 2,
           let value = Int(match[1] ?? "") {
            leave = value
            if let range = tail.range(of: #"to\s+last\s+\d+\s*sts?"#,
                                      options: [.regularExpression, .caseInsensitive]) {
                suffixText = String(tail[range.upperBound...])
            }
        } else if let range = tail.range(of: #"to\s+end"#,
                                         options: [.regularExpression, .caseInsensitive]) {
            suffixText = String(tail[range.upperBound...])
        }

        return StarRepeat(
            prefix: tokens(prefixText),
            body: tokens(bodyText),
            suffix: tokens(suffixText),
            leaveStitches: leave)
    }

    struct Flexible {
        /// Instructions worked before the "to end" / "to last N sts" marker.
        var before: [String]
        /// Instructions worked after it — these are the "last N stitches".
        var after: [String]
        var hasToEnd: Bool
        var leaveStitches: Int

        var tokens: [String] { before + after }
    }

    /// Handles "knit to end" and "knit to last 3 sts" outside a star repeat.
    static func flexibleConsumption(_ text: String) -> Flexible {
        var hasToEnd = false
        var leave = 0
        var before: [String] = []
        var after: [String] = []

        for token in tokens(text) {
            if let match = capture(#"^(?:k|p|knit|purl)\s*to\s*last\s*(\d+)\s*sts?$"#, in: token),
               match.count >= 2, let value = Int(match[1] ?? "") {
                hasToEnd = true
                leave = value
                continue
            }
            if token.range(of: #"^(?:k|p|knit|purl|work)\s*to\s*(?:end|marker)$"#,
                           options: [.regularExpression]) != nil {
                hasToEnd = true
                continue
            }
            if hasToEnd {
                after.append(token)
            } else {
                before.append(token)
            }
        }
        return Flexible(before: before, after: after, hasToEnd: hasToEnd, leaveStitches: leave)
    }

    static func tokens(_ text: String) -> [String] {
        text
            .replacingOccurrences(of: ";", with: ",")
            .components(separatedBy: ",")
            .map {
                $0.trimmingCharacters(in: CharacterSet(charactersIn: " .;:*()[]-"))
            }
            .filter { !$0.isEmpty }
            // Drop trailing "(96 sts)" style annotations.
            .filter { capture(#"^\d+\s*sts?$"#, in: $0) == nil }
    }

    // MARK: - Regex helper

    /// Returns the whole match plus each capture group, or nil if no match.
    static func capture(_ pattern: String, in text: String) -> [String?]? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
        else { return nil }
        let range = NSRange(text.startIndex ..< text.endIndex, in: text)
        guard let match = regex.firstMatch(in: text, options: [], range: range) else { return nil }
        return (0 ..< match.numberOfRanges).map { index in
            guard let groupRange = Range(match.range(at: index), in: text) else { return nil }
            return String(text[groupRange])
        }
    }

    static func number(_ value: String?) -> Double? {
        guard let value else { return nil }
        return Double(value.replacingOccurrences(of: ",", with: "."))
    }
}

File: android/engine/src/main/kotlin/com/knitstudio/engine/WrittenPatternParser.kt
package com.knitstudio.engine

/** One instruction inside a row, and what it does to the stitch count. */
data class StitchOperation(val text: String, val consumes: Int, val produces: Int)

/** One row of a written pattern, after parsing. */
data class ParsedRow(
    val label: String,
    val side: String? = null,
    val text: String,
    val startingStitches: Int? = null,
    val endingStitches: Int? = null,
    val issues: List<String> = emptyList(),
) {
    val isVerified: Boolean get() = endingStitches != null && issues.isEmpty()
}

/** The result of importing a written pattern. */
data class ParsedPattern(
    val title: String? = null,
    val gauge: Gauge? = null,
    val gaugeNote: String? = null,
    val castOnStitches: Int? = null,
    val yarnNotes: List<String> = emptyList(),
    val needleNotes: List<String> = emptyList(),
    val rows: List<ParsedRow> = emptyList(),
    val unknownTerms: List<String> = emptyList(),
    val issues: List<String> = emptyList(),
) {
    val verifiedRowCount: Int get() = rows.count { it.isVerified }
}

/**
 * Reads a written knitting pattern and works out what each row does to the
 * stitch count — which is how you find the row where a pattern (or a
 * transcription of it) stops adding up.
 */
object WrittenPatternParser {

    data class Outcome(
        val produced: Int? = null,
        val issues: List<String> = emptyList(),
        val unknown: List<String> = emptyList(),
    )

    private data class Tally(val consumes: Int, val produces: Int)

    private data class StarRepeat(
        val prefix: List<String>,
        val body: List<String>,
        val suffix: List<String>,
        val leaveStitches: Int,
    )

    private data class Flexible(
        val before: List<String>,
        val after: List<String>,
        val hasToEnd: Boolean,
        val leaveStitches: Int,
    ) {
        val tokens: List<String> get() = before + after
    }

    // MARK: - Entry point

    fun parse(raw: String): ParsedPattern {
        val lines = raw.replace("\r\n", "\n").split("\n").map { it.trim() }

        var title: String? = lines.firstOrNull { it.isNotEmpty() && it.length < 80 }
        var gauge: Gauge? = null
        var gaugeNote: String? = null
        var castOn: Int? = null
        val yarnNotes = mutableListOf<String>()
        val needleNotes = mutableListOf<String>()
        val rows = mutableListOf<ParsedRow>()

        for (line in lines) {
            if (line.isEmpty()) continue
            val lower = line.lowercase()

            if (gauge == null &&
                (lower.contains("gauge") || lower.contains("tension") ||
                    (lower.contains("sts") && (lower.contains("row") || lower.contains("rnd"))))
            ) {
                parseGauge(line)?.let { gauge = it; gaugeNote = line }
            }
            if (lower.contains("yarn") || lower.contains("skein") || lower.contains("ball") ||
                lower.contains("yardage") || lower.contains("metres") || lower.contains("meters")
            ) {
                yarnNotes += line
            }
            if (lower.contains("needle") || (lower.contains("mm") && lower.contains("circular"))) {
                needleNotes += line
            }
            if (castOn == null) parseCastOn(line)?.let { castOn = it }
            parseRowLine(line)?.let { rows += it }
        }

        // Walk the rows, threading the stitch count through.
        var running = castOn
        val unknown = linkedSetOf<String>()
        val resolved = rows.map { row ->
            val start = running
            val outcome = evaluate(row.text, start)
            unknown += outcome.unknown
            running = outcome.produced ?: running
            row.copy(
                startingStitches = start,
                endingStitches = outcome.produced,
                issues = outcome.issues,
            )
        }

        val issues = mutableListOf<String>()
        if (resolved.isEmpty()) {
            issues += "No numbered rows found. This importer reads patterns written as " +
                "‘Row 1 (RS): k2, p2…’ — a scanned image of a chart will not parse."
        }
        if (castOn == null && resolved.isNotEmpty()) {
            issues += "No cast-on count found, so stitch counts could not be checked. " +
                "Add a line like ‘Cast on 96 sts’ and import again."
        }

        return ParsedPattern(
            title = title,
            gauge = gauge,
            gaugeNote = gaugeNote,
            castOnStitches = castOn,
            yarnNotes = yarnNotes,
            needleNotes = needleNotes,
            rows = resolved,
            unknownTerms = unknown.sorted(),
            issues = issues,
        )
    }

    // MARK: - Header lines

    fun parseGauge(line: String): Gauge? {
        val match = capture(
            """(\d+(?:[.,]\d+)?)\s*(?:sts?|stitches)\D{0,40}?(\d+(?:[.,]\d+)?)\s*(?:rows?|rnds?|rounds?)""",
            line,
        ) ?: return null
        val stitches = number(match.getOrNull(1)) ?: return null
        val rows = number(match.getOrNull(2)) ?: return null

        // Work out whether the gauge is quoted over 10 cm or 4 inches.
        var window = 10.0
        val unit = capture(
            """(\d+(?:[.,]\d+)?)\s*(cm|centimetre|centimeter|in\b|inch|inches|")""", line,
        )
        if (unit != null) {
            val value = number(unit.getOrNull(1))
            val label = unit.getOrNull(2)?.lowercase() ?: ""
            if (value != null) {
                window = if (label.startsWith("in") || label == "\"") value * 2.54 else value
            }
        } else if (line.lowercase().contains("4\"") || line.lowercase().contains("4 in")) {
            window = 10.16
        }
        if (window <= 0) return null
        return Gauge.fromSwatch(stitches, rows, window)
    }

    fun parseCastOn(line: String): Int? {
        // The \b stops "co" matching inside another word.
        val match = capture("""(?:cast on|\bco)\s+(\d+)\s*(?:sts?|stitches)?""", line) ?: return null
        return number(match.getOrNull(1))?.toInt()
    }

    fun parseRowLine(line: String): ParsedRow? {
        val match = capture(
            """^(rows?|rnds?|rounds?)\s*([\d\s,\-–and]+?)\s*(?:\((rs|ws)\))?\s*[:.]\s*(.+)$""",
            line,
        ) ?: return null

        val keyword = (match.getOrNull(1) ?: "Row").replaceFirstChar { it.uppercase() }
        val numbers = (match.getOrNull(2) ?: "").trim()
        val side = match.getOrNull(3)?.uppercase()
        val body = (match.getOrNull(4) ?: "").trim()
        if (body.isEmpty()) return null

        return ParsedRow(label = "$keyword $numbers", side = side, text = body)
    }

    // MARK: - Row evaluation

    /** Works out the stitch count a row ends on, given what it started with. */
    fun evaluate(text: String, incoming: Int?): Outcome {
        val issues = mutableListOf<String>()
        val unknown = mutableListOf<String>()
        val normalised = normalise(text)
        val stated = statedCount(normalised)

        if (incoming == null) return Outcome(produced = stated)

        val expanded = expandBrackets(normalised)

        val star = splitStarRepeat(expanded)
        if (star != null) {
            val prefix = tally(star.prefix, unknown)
            val body = tally(star.body, unknown)
            val suffix = tally(star.suffix, unknown)

            if (body.consumes <= 0) {
                issues += "The repeated section does not use any stitches, so it cannot be counted."
                return Outcome(stated, issues, unknown)
            }
            // "to last N sts" describes the same stitches the suffix works, so
            // it is a consistency check, not an extra deduction.
            if (star.leaveStitches > 0 && suffix.consumes != star.leaveStitches) {
                issues += "‘to last ${star.leaveStitches} sts’ but the instructions after it use " +
                    "${suffix.consumes} sts."
            }
            val available = incoming - prefix.consumes - suffix.consumes
            if (available < 0) {
                issues += "This row needs more stitches than the $incoming on the needle."
                return Outcome(stated, issues, unknown)
            }
            val repeats = available / body.consumes
            if (available % body.consumes != 0) {
                issues += "The repeat does not divide evenly: $available sts left for a " +
                    "${body.consumes}-st repeat leaves ${available % body.consumes} over."
            }
            val produced = prefix.produces + repeats * body.produces + suffix.produces
            checkStated(produced, stated, issues)
            return Outcome(produced, issues, unknown)
        }

        val flexible = flexibleConsumption(expanded)
        val fixed = tally(flexible.tokens, unknown)
        if (flexible.hasToEnd) {
            // The instructions after "to last N sts" are exactly those N sts —
            // anything before the marker is not part of that count.
            val after = tally(flexible.after, mutableListOf())
            if (flexible.leaveStitches > 0 && after.consumes != flexible.leaveStitches) {
                issues += "‘to last ${flexible.leaveStitches} sts’ but the instructions after it use " +
                    "${after.consumes} sts."
            }
            val available = incoming - fixed.consumes
            if (available < 0) {
                issues += "This row needs more stitches than the $incoming on the needle."
                return Outcome(stated, issues, unknown)
            }
            val produced = fixed.produces + available
            checkStated(produced, stated, issues)
            return Outcome(produced, issues, unknown)
        }

        if (fixed.consumes != incoming) {
            issues += "This row uses ${fixed.consumes} sts but $incoming were on the needle."
        }
        checkStated(fixed.produces, stated, issues)
        return Outcome(fixed.produces, issues, unknown)
    }

    private fun checkStated(produced: Int, stated: Int?, issues: MutableList<String>) {
        if (stated != null && stated != produced) {
            issues += "The pattern says $stated sts at the end of this row; " +
                "the instructions give $produced."
        }
    }

    // MARK: - Tokenising

    private fun tally(tokens: List<String>, unknown: MutableList<String>): Tally {
        var consumes = 0
        var produces = 0
        for (token in tokens) {
            val operation = interpret(token)
            if (operation == null) {
                if (token.isNotEmpty()) unknown += token
                continue
            }
            consumes += operation.consumes
            produces += operation.produces
        }
        return Tally(consumes, produces)
    }

    private val noOps = listOf("pm", "sm", "place marker", "slip marker", "remove marker", "turn", "rm")

    private val fixedOperations: List<Triple<String, Int, Int>> = listOf(
        Triple("sl2-k1-p2sso", 3, 1), Triple("sl2 k1 p2sso", 3, 1),
        Triple("s2kp", 3, 1), Triple("sk2p", 3, 1),
        Triple("cdd", 3, 1), Triple("k3tog", 3, 1), Triple("p3tog", 3, 1),
        Triple("k2tog tbl", 2, 1), Triple("p2tog tbl", 2, 1),
        Triple("k2tog", 2, 1), Triple("p2tog", 2, 1), Triple("ssk", 2, 1), Triple("ssp", 2, 1),
        Triple("kfb", 1, 2), Triple("pfb", 1, 2),
        Triple("m1lp", 0, 1), Triple("m1rp", 0, 1), Triple("m1l", 0, 1), Triple("m1r", 0, 1),
        Triple("m1p", 0, 1), Triple("m1", 0, 1),
        Triple("yo", 0, 1), Triple("yon", 0, 1), Triple("yfwd", 0, 1), Triple("yrn", 0, 1),
    )

    /** Turns one instruction into its stitch arithmetic. */
    fun interpret(token: String): StitchOperation? {
        val text = token.trim()
        if (text.isEmpty()) return null

        for (noop in noOps) {
            if (text == noop || text.startsWith("$noop ")) return StitchOperation(text, 0, 0)
        }
        for ((name, consumes, produces) in fixedOperations) {
            if (text == name || text.startsWith("$name ")) return StitchOperation(text, consumes, produces)
        }

        capture("""^(k|p|knit|purl|sl|slip)\s*(\d+)?(?:\s*sts?)?(?:\s*tbl)?$""", text)?.let { match ->
            val count = match.getOrNull(2)?.toIntOrNull() ?: 1
            return StitchOperation(text, count, count)
        }
        capture("""^(?:bind off|bo|cast off)\s*(\d+)""", text)?.let { match ->
            match.getOrNull(1)?.toIntOrNull()?.let { return StitchOperation(text, it, 0) }
        }
        capture("""^(?:cast on|co)\s*(\d+)""", text)?.let { match ->
            match.getOrNull(1)?.toIntOrNull()?.let { return StitchOperation(text, 0, it) }
        }
        return null
    }

    // MARK: - Structure

    fun normalise(text: String): String = text.lowercase()
        .replace("–", "-").replace("—", "-").replace("×", "x")
        .replace("  ", " ")
        .trim()

    /** Reads a trailing "(96 sts)" or "- 96 sts" the pattern states for itself. */
    fun statedCount(text: String): Int? =
        capture("""[(\-]\s*(\d+)\s*sts?\s*\)?\s*$""", text)?.getOrNull(1)?.toIntOrNull()

    /** Expands "[k2, p2] 3 times" and "(k1, yo) x4" into plain instructions. */
    fun expandBrackets(text: String): String {
        var working = text
        val regex = Regex("""[\[(]([^\])]+)[\])]\s*(?:x\s*)?(\d+)\s*(?:times?)?""", RegexOption.IGNORE_CASE)
        // Bounded so a malformed pattern cannot loop forever.
        repeat(8) {
            val match = regex.find(working) ?: return working
            val count = match.groupValues[2].toIntOrNull() ?: return working
            if (count <= 0 || count >= 400) return working
            val body = match.groupValues[1].trim()
            val expansion = List(count) { body }.joinToString(", ")
            working = working.substring(0, match.range.first) + expansion +
                working.substring(match.range.last + 1)
        }
        return working
    }

    /** Splits "k3, *p2, k2; rep from * to last 3 sts, k3" into its three parts. */
    private fun splitStarRepeat(text: String): StarRepeat? {
        val star = text.indexOf('*')
        if (star < 0) return null
        val prefixText = text.substring(0, star)
        val rest = text.substring(star + 1)

        val repeatMatch = Regex("""rep(eat)?\s+from\s*\*""", RegexOption.IGNORE_CASE).find(rest)
            ?: return null
        val bodyText = rest.substring(0, repeatMatch.range.first)
        val tail = rest.substring(repeatMatch.range.last + 1)

        var leave = 0
        var suffixText = tail
        val lastMatch = Regex("""to\s+last\s+(\d+)\s*sts?""", RegexOption.IGNORE_CASE).find(tail)
        if (lastMatch != null) {
            leave = lastMatch.groupValues[1].toIntOrNull() ?: 0
            suffixText = tail.substring(lastMatch.range.last + 1)
        } else {
            Regex("""to\s+end""", RegexOption.IGNORE_CASE).find(tail)?.let {
                suffixText = tail.substring(it.range.last + 1)
            }
        }

        return StarRepeat(tokens(prefixText), tokens(bodyText), tokens(suffixText), leave)
    }

    /** Handles "knit to end" and "knit to last 3 sts" outside a star repeat. */
    private fun flexibleConsumption(text: String): Flexible {
        var hasToEnd = false
        var leave = 0
        val before = mutableListOf<String>()
        val after = mutableListOf<String>()

        for (token in tokens(text)) {
            val lastMatch = capture("""^(?:k|p|knit|purl)\s*to\s*last\s*(\d+)\s*sts?$""", token)
            if (lastMatch != null) {
                hasToEnd = true
                leave = lastMatch.getOrNull(1)?.toIntOrNull() ?: 0
                continue
            }
            if (capture("""^(?:k|p|knit|purl|work)\s*to\s*(?:end|marker)$""", token) != null) {
                hasToEnd = true
                continue
            }
            if (hasToEnd) after += token else before += token
        }
        return Flexible(before, after, hasToEnd, leave)
    }

    fun tokens(text: String): List<String> = text
        .replace(";", ",")
        .split(",")
        .map { it.trim(' ', '.', ';', ':', '*', '(', ')', '[', ']', '-') }
        .filter { it.isNotEmpty() }
        // Drop trailing "(96 sts)" style annotations.
        .filter { capture("""^\d+\s*sts?$""", it) == null }

    // MARK: - Regex helper

    /** Returns the whole match plus each capture group, or null if no match. */
    private fun capture(pattern: String, text: String): List<String?>? {
        val match = Regex(pattern, RegexOption.IGNORE_CASE).find(text) ?: return null
        return (0 until match.groups.size).map { match.groups[it]?.value }
    }

    private fun number(value: String?): Double? =
        value?.replace(",", ".")?.toDoubleOrNull()
}
